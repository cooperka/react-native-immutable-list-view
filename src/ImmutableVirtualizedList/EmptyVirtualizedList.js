import Immutable from 'immutable';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import { Text, VirtualizedList } from 'react-native';

import ImmutableVirtualizedList from './ImmutableVirtualizedList';

import styles from '../styles';
import utils from '../utils';

/**
 * A VirtualizedList that displays a single item showing that there is nothing to display.
 * Useful e.g. for preserving the ability to pull-refresh an empty list.
 */
class EmptyVirtualizedList extends PureComponent {

  static propTypes = {
    // Pass through any props that VirtualizedList would normally take.
    ...VirtualizedList.propTypes,

    // Make this prop optional instead of required.
    renderItem: PropTypes.func,

    emptyText: PropTypes.string,
  };

  static defaultProps = {
    ...VirtualizedList.defaultProps,

    keyExtractor: (_, index) => String(index),

    emptyText: 'No data.',
  };

  state = {
    listData: utils.UNITARY_LIST,
  };

  componentWillMount() {
    this.setListDataFromProps(this.props);
  }

  componentWillReceiveProps(nextProps) {
    this.setListDataFromProps(nextProps);
  }

  setListDataFromProps(props) {
    const { listData } = this.state;
    const { renderEmpty, renderEmptyInList, emptyText } = props;

    // Update the data to make sure the list re-renders if any of the relevant props have changed.
    this.setState({
      listData: listData.set(0, Immutable.fromJS([renderEmpty, renderEmptyInList, emptyText])),
    });
  }

  /**
   * Returns a simple text element showing the `emptyText` string.
   * This method can be overridden by passing in your own `renderItem` prop instead.
   */
  renderItem() {
    const { emptyText } = this.props;

    return (
      <Text style={styles.emptyText}>
        {emptyText}
      </Text>
    );
  }

  render() {
    const { listData } = this.state;
    const { renderEmpty, renderEmptyInList, ...passThroughProps } = this.props;

    return (
      <ImmutableVirtualizedList
        renderItem={() => this.renderItem()}
        {...passThroughProps}
        immutableData={listData}
      />
    );
  }

}

export { EmptyVirtualizedList };
