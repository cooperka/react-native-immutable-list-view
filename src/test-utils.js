/* eslint-disable import/no-extraneous-dependencies */
import Immutable from 'immutable';
import React from 'react';
import { Text, ListView } from 'react-native';
import renderer from 'react-test-renderer';
/* eslint-enable */

import ImmutableListView from './ImmutableListView';
import ImmutableVirtualizedList from './ImmutableVirtualizedList';

/**
 * Some common types of data you may want to render with ImmutableListView.
 * @see https://facebook.github.io/react-native/docs/listviewdatasource.html#constructor
 */
const data = {
  EMPTY_DATA: Immutable.List(),

  LIST_DATA: Immutable.List(['lists', 'are', 'great']),

  LIST_DATA_NESTED: Immutable.List([['so', 'are'], ['nested', 'lists']]),

  MAP_DATA_LIST_ROWS: Immutable.fromJS({
    first: ['m', 'a', 'p'],
    second: ['foo'],
    third: [],
    fourth: ['bar'],
  }),

  MAP_DATA_MAP_ROWS: Immutable.fromJS({
    first: {
      row1: 'data 1',
      row2: 'data 2',
    },
    second: {},
  }),

  SET_DATA: Immutable.Set(['one', 'two', 'three']),

  RANGE_DATA: Immutable.Range(3, 10, 3),
};

const renderers = {
  /**
   * @param {*} rowData
   */
  renderRow(rowData) {
    return <Text>{JSON.stringify(rowData)}</Text>;
  },

  // eslint-disable-next-line react/prop-types
  renderItem({ item }) {
    return <Text>{JSON.stringify(item)}</Text>;
  },

  /**
   * @param {Immutable.Iterable} sectionData
   * @param {String} category
   */
  renderSectionHeader(sectionData, category) {
    return <Text header>{`${category} (${sectionData.size} items)`}</Text>;
  },
};

const mocks = {
  /**
   * Mock ScrollView so that it doesn't contain any props when rendered by ListView.
   * This is useful for comparison between ListView and ImmutableListView.
   *
   * @returns {ImmutableListView}
   */
  getImmutableListViewWithoutProps() {
    jest.resetModules();

    // eslint-disable-next-line react/prop-types
    const mockScrollView = props =>
      React.createElement('ScrollView', {}, props.children);
    jest.doMock('ScrollView', () => mockScrollView);

    // eslint-disable-next-line global-require
    return require('./ImmutableListView').default;
  },
};

const expectors = {
  expectToMatchSnapshotWithData(immutableData, shouldRenderSectionHeaders) {
    const renderSectionHeaderProps = shouldRenderSectionHeaders
      ? { renderSectionHeader: renderers.renderSectionHeader }
      : {};

    const tree = renderer
      .create(
        <ImmutableListView
          immutableData={immutableData}
          renderRow={renderers.renderRow}
          {...renderSectionHeaderProps}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  },

  expectVirtualizedToMatchSnapshotWithData(immutableData) {
    const tree = renderer
      .create(
        <ImmutableVirtualizedList
          immutableData={immutableData}
          renderItem={renderers.renderItem}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  },

  expectToMatchListViewWithData(immutableData, shouldRenderSectionHeaders) {
    const MockedImmutableListView = mocks.getImmutableListViewWithoutProps();

    const dataSource = new ListView.DataSource({
      rowHasChanged: (r1, r2) => r1 !== r2,
      sectionHeaderHasChanged: (s1, s2) => s1 !== s2,
    });

    const renderSectionHeaderProps = shouldRenderSectionHeaders
      ? { renderSectionHeader: renderers.renderSectionHeader }
      : {};

    const immutableTree = renderer
      .create(
        <MockedImmutableListView
          immutableData={immutableData}
          renderRow={renderers.renderRow}
          renderEmptyInList={null}
          {...renderSectionHeaderProps}
        />
      )
      .toJSON();

    const updatedDataSource = dataSource.cloneWithRows(immutableData.toJS());
    const regularTree = renderer
      .create(
        <ListView
          dataSource={updatedDataSource}
          renderRow={renderers.renderRow}
          {...renderSectionHeaderProps}
        />
      )
      .toJSON();

    expect(immutableTree).toEqual(regularTree);
  },
};

export { data, renderers, mocks, expectors };
