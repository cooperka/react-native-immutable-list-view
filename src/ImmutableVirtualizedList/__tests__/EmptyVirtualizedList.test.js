import React from 'react';
import renderer from 'react-test-renderer';

import { renderers } from '../../test-utils';

import { EmptyVirtualizedList } from '../EmptyVirtualizedList';

describe('EmptyVirtualizedList', () => {
  it('renders with default text', () => {
    const tree = renderer.create(<EmptyVirtualizedList />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders with custom text', () => {
    const tree = renderer.create(
      <EmptyVirtualizedList emptyText="Nothing. Nothing at all." />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders with custom renderRow', () => {
    const tree = renderer.create(
      <EmptyVirtualizedList
        emptyText="Nothing. Nothing at all."
        renderItem={() => renderers.renderRow('Overridden!')}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
