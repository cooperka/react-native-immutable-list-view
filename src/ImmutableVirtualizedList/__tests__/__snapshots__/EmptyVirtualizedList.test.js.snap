// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmptyVirtualizedList renders with custom renderRow 1`] = `
<RCTScrollView
  data={
    Immutable.List [
      Immutable.List [
        undefined,
        undefined,
        "Nothing. Nothing at all.",
      ],
    ]
  }
  disableVirtualization={false}
  emptyText="Nothing. Nothing at all."
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  immutableData={
    Immutable.List [
      Immutable.List [
        undefined,
        undefined,
        "Nothing. Nothing at all.",
      ],
    ]
  }
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderEmptyInList="No data."
  renderItem={[Function]}
  renderScrollComponent={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
      >
        "Overridden!"
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`EmptyVirtualizedList renders with custom text 1`] = `
<RCTScrollView
  data={
    Immutable.List [
      Immutable.List [
        undefined,
        undefined,
        "Nothing. Nothing at all.",
      ],
    ]
  }
  disableVirtualization={false}
  emptyText="Nothing. Nothing at all."
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  immutableData={
    Immutable.List [
      Immutable.List [
        undefined,
        undefined,
        "Nothing. Nothing at all.",
      ],
    ]
  }
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderEmptyInList="No data."
  renderItem={[Function]}
  renderScrollComponent={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "padding": 8,
            "textAlign": "center",
          }
        }
      >
        Nothing. Nothing at all.
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`EmptyVirtualizedList renders with default text 1`] = `
<RCTScrollView
  data={
    Immutable.List [
      Immutable.List [
        undefined,
        undefined,
        "No data.",
      ],
    ]
  }
  disableVirtualization={false}
  emptyText="No data."
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  immutableData={
    Immutable.List [
      Immutable.List [
        undefined,
        undefined,
        "No data.",
      ],
    ]
  }
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderEmptyInList="No data."
  renderItem={[Function]}
  renderScrollComponent={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "padding": 8,
            "textAlign": "center",
          }
        }
      >
        No data.
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
